### PR TITLE
Jetty 9.4.x 2550 coalesce ranges

### PR DESCRIPTION
**Pull request by gregw**
For #2550, coalesce byte ranges.
written pair programming with @lachlan-roberts based on tests/cleanup from @olamy

### CODE REVIEW

File: jetty-server/src/test/java/org/eclipse/jetty/server/InclusiveByteRangeTest.java [hunk 1]
```diff
@@ -124,42 +138,116 @@ public void testMultipleAbsoluteRanges()
...
+
+        rangeString = "bytes=90-100, 10-20, 30-40, -161";
+        List<InclusiveByteRange> ranges = parseRanges(size,rangeString);
+        
+        assertEquals("Satisfiable Ranges of [" + rangeString + "] count",2,ranges.size());
+        Iterator<InclusiveByteRange> inclusiveByteRangeIterator = ranges.iterator();
+        assertRange("Range [" + rangeString + "]",30,199,size,inclusiveByteRangeIterator.next());
+        assertRange("Range [" + rangeString + "]",10,20,size,inclusiveByteRangeIterator.next());
+    }
+
```
Comment (joakime) [inline jetty-server/src/test/java/org/eclipse/jetty/server/InclusiveByteRangeTest.java]: Should probably test other bad range definitions.

eg:

```
bytes=1-50,1-b,a-50
bytes=a-b
bytes=50-50,100-100,121-121
bytes=01-050
bytes=0F-FF
bytes=1--50
bytes= 1 - 50
bytes=\t1\t-\t50
bytes=1-50;90-101;200-250
bytes=50-1
bytes=050-001
bytes= 50-
```

File: jetty-server/src/main/java/org/eclipse/jetty/server/InclusiveByteRange.java [hunk 2]
```diff
@@ -100,52 +169,89 @@ public long getLast()
...
+                            last = Long.parseLong(t.substring(dash + 1).trim());
+                        
+                        if (first==-1)
                         {
-                            if (d + 1 < t.length())
-                                last = Long.parseLong(t.substring(d + 1).trim());
-                            else
+                            if (last==-1)
                             {
                                 LOG.warn("Bad range format: {}",t);
```
Comment (joakime) [inline jetty-server/src/main/java/org/eclipse/jetty/server/InclusiveByteRange.java]: I'm beginning to think that and bad range should result in a 400 Bad Request response.

File: jetty-server/src/test/java/org/eclipse/jetty/server/InclusiveByteRangeTest.java [hunk 3]
```diff
@@ -124,48 +138,201 @@ public void testMultipleAbsoluteRanges()
...
         assertSimpleRange(5,10,"bytes=5-10",200);
         assertSimpleRange(195,199,"bytes=-5",200);
         assertSimpleRange(50,119,"bytes=50-150",120);
         assertSimpleRange(50,119,"bytes=50-",120);
+        
+        assertSimpleRange(1,50,"bytes= 1 - 50",120);
+    }
+
+    // TODO: evaluate this vs assertInvalidRange() above, which behavior is correct? null? or empty list?
+    private void assertBadRangeList(int size, String badRange)
```
Comment (joakime) [inline jetty-server/src/test/java/org/eclipse/jetty/server/InclusiveByteRangeTest.java]: Yeah, this should be merged up with assertInvalidRange() above.  My mistake.

File: jetty-server/src/test/java/org/eclipse/jetty/server/InclusiveByteRangeTest.java [hunk 4]
```diff
@@ -124,48 +138,201 @@ public void testMultipleAbsoluteRanges()
...
+    
+        List<InclusiveByteRange> ranges = InclusiveByteRange.satisfiableRanges(strings.elements(),size);
+        // if one part is bad, the entire set of ranges should be treated as bad, per RFC7233
+        assertThat("Should have no ranges", ranges, is(nullValue()));
+    }
+    
+    @Test
+    public void testBadRange_SetPartiallyBad()
+    {
+        assertBadRangeList(500, "bytes=1-50,1-b,a-50");
```
Comment (joakime) [inline jetty-server/src/test/java/org/eclipse/jetty/server/InclusiveByteRangeTest.java]: This is an example of a partially satisfiable range (1-50) with some bad ranges.
This is the kind of request that RFC7233 seems to indicate should result in a 400 Bad Request.

File: jetty-server/src/test/java/org/eclipse/jetty/server/InclusiveByteRangeTest.java [hunk 5]
```diff
@@ -124,48 +138,201 @@ public void testMultipleAbsoluteRanges()
...
+    public void testBadRange_NegativeSize()
+    {
+        assertBadRangeList(500, "bytes=50-1");
+    }
+    
+    @Test
+    public void testBadRange_ZeroedNegativeSize()
+    {
+        assertBadRangeList(500, "bytes=050-001");
     }
```
Comment (joakime) [inline jetty-server/src/test/java/org/eclipse/jetty/server/InclusiveByteRangeTest.java]: Forgot to add tests for ...

```
bytes=1--50    // seen as [1] to [-50]
bytes=1---50   // seen as [1] to [--50]
```

File: jetty-server/src/main/java/org/eclipse/jetty/server/InclusiveByteRange.java (hunk 2 above)
Comment (gregw) [reply jetty-server/src/main/java/org/eclipse/jetty/server/InclusiveByteRange.java]: Probably, but let's not change that behaviour in this release.... we'd have to add compliance modes etc.

**Pull request merged**